### ISSUE TITLE
Strange behavior when explicit annotation is given to AsyncGenerator

### ISSUE THREAD

**Issue by kyle-query** [microsoft/pyright#5411]
**Describe the bug**

This could be user error, but I remove type annotations, pyright correctly infers the type and doesn't show any type errors. However if I use the same annotation that pyright inferred, it reports a type error in the function body.

**To Reproduce**

```python
from typing import Any, Generator

async def test0(): # <-- inferred: def test0() -> Generator[int, Any, None]
    xs = [1, 2, 3]
    for x in xs:
        yield x

async def test1() -> Generator[int, Any, None]:
    xs = [1, 3, 5]
    for x in xs:
        yield x # <-- error

# Sanity check, they are assigned the same type
test0 # inferred: def test0() -> AsyncGenerator[int, Any]
test1 # inferred: def test1() -> AsyncGenerator[int, Any]
```

The type error is:

```
Return type of async generator function must be compatible with "AsyncGenerator[dict[str, int], Any]"
  "AsyncGenerator[dict[str, int]]" is incompatible with "Generator[dict[str, int], None, None]"
```

**Expected behavior**

I don't expect a type error. Even if the type error is correct, I would expect both functions to have identical type analyses because they have identical bodies and type assignments.

**VS Code extension or command-line**

I can reproduce this in neovim using pyright 1.1.316; and in Visual Studio Code with pylance v2023.6.42.

**Additional context**

 I have the following configuration in pyproject.toml:

```toml
[tool.pyright]
pythonVersion                       = "3.10"
pythonPlatform                      = "Linux"
typeCheckingMode                    = "strict"
reportImportCycles                  = "information"
reportUnusedImport                  = "warning"
reportUnusedClass                   = "warning"
reportUnusedFunction                = "warning"
reportUnusedVariable                = "warning"
reportMissingTypeStubs              = "warning"
reportCallInDefaultInitializer      = "error"
reportImplicitStringConcatenation   = "error"
reportUnnecessaryTypeIgnoreComment  = "error"
```

**Comment by rsokl** [comment]
The inferred return type is `AsyncGenerator`, not `Generator`:

```python
from typing import Any, AsyncGenerator

async def test0():
    xs = [1, 2, 3]
    for x in xs:
        yield x

reveal_type(test0)  # reveals: Generator[int, Any, None]

async def test1() -> AsyncGenerator[int, Any]:
    xs = [1, 3, 5]
    for x in xs:
        yield x # no error
```

If you update your code accordingly you will not see an error. I think the issue is that if you cursor-hover over `test0` in its definition, you see `Generator[int, Any, None]`, but that is not the actual inferred type, as demonstrated by the `reveal_type` call (which can be used to prompt a type-checker to reveal its inferred type).

**Comment by kyle-query** [comment]
Thanks, that indeed has the result I want. I found this counterintuitive because [here](https://github.com/microsoft/pyright/issues/5080#issuecomment-1539280043) the author explains `async` changes the return type. Maybe this is a special case where it doesn't.

**Comment by erictraut** [comment]
I agree that the current behavior is inconsistent. I've addressed this so the hover text will display `AsyncGenerator` rather than `Generator` in this case. This will be in the next release.
